doc/man3/STORE_LOADER.pod

```diff
@@ -0,0 +1,202 @@
+=pod
+
+=head1 NAME
+
+STORE_LOADER, STORE_LOADER_CTX, STORE_LOADER_new,
+STORE_LOADER_set0_scheme, STORE_LOADER_set_open,
+STORE_LOADER_set_ctrl, STORE_LOADER_set_load, STORE_LOADER_set_eof,
+STORE_LOADER_set_close, STORE_LOADER_free, STORE_register_loader,
+STORE_unregister_loader, STORE_open_fn, STORE_ctrl_fn, STORE_load_fn,
+STORE_eof_fn, STORE_close_fn - Types and functions to manipulate,
+register and unregister STORE loaders for different URI schemes 
+
+=head1 SYNOPSIS
+
+ #include <openssl/store.h>
+
+ typedef struct store_loader_st STORE_LOADER;
+
+ STORE_LOADER *STORE_LOADER_new(void);
+ int STORE_LOADER_set0_scheme(STORE_LOADER *store_loader, const char *scheme);
+
+ /* struct store_loader_ctx_st is defined differently by each loader */
+ typedef struct store_loader_ctx_st STORE_LOADER_CTX;
+
+ typedef STORE_LOADER_CTX *(*STORE_open_fn)(const char *uri,
+                                            const UI_METHOD *ui_method,
+                                            void *ui_data);
+ int STORE_LOADER_set_open(STORE_LOADER *store_loader,
+                           STORE_open_fn store_open_function);
+ typedef int (*STORE_ctrl_fn)(STORE_LOADER_CTX *ctx, int cmd, va_list args);
+ int STORE_LOADER_set_ctrl(STORE_LOADER *store_loader,
+                           STORE_ctrl_fn store_ctrl_function);
+ typedef STORE_INFO *(*STORE_load_fn)(STORE_LOADER_CTX *ctx,
+                                      UI_METHOD *ui_method, void *ui_data);
+ int STORE_LOADER_set_load(STORE_LOADER *store_loader,
+                           STORE_load_fn store_load_function);
+ typedef int (*STORE_eof_fn)(STORE_LOADER_CTX *ctx);
+ int STORE_LOADER_set_eof(STORE_LOADER *store_loader,
+                            STORE_eof_fn store_eof_function);
+ typedef int (*STORE_close_fn)(STORE_LOADER_CTX *ctx);
+ int STORE_LOADER_set_close(STORE_LOADER *store_loader,
+                            STORE_close_fn store_close_function);
+ void STORE_LOADER_free(STORE_LOADER *store_loader);
+ 
+ int STORE_register_loader(STORE_LOADER *loader);
+ STORE_LOADER *STORE_unregister_loader(const char *scheme);
+
+=head1 DESCRIPTION
+
+These functions help applications and engines to create loaders for
+schemes they support.
+
+=head2 Types
+
+B<STORE_LOADER> is the type to hold a loader.  It contains a scheme
+and the functions needed to implement STORE_open(), STORE_eof(),
+STORE_load() and STORE_close() for this scheme.
+
+B<STORE_LOADER_CTX> is a type template, to be defined by each loader
+using B<struct store_loader_ctx_st { ... }>.
```

Comment:
:scream:
Hmmm.....right...now I understand. I really don't like that. Lots of different definitions of struct store_loader_ctx_st depending on what translation unit you happen to be in?? It also of course forces the application to use a new translation unit for each store_loader_ctx_st, which means we are requiring a user to structure their code in a particular way which may not be appropriate. Why can't we just make this a void? i.e.

````
--- a/include/openssl/store.h
+++ b/include/openssl/store.h
@@ -161,7 +161,7 @@ STORE_LOADER *STORE_LOADER_new(ENGINE *e, const char *scheme);
 const ENGINE *STORE_LOADER_get0_engine(const STORE_LOADER *store_loader);
 const char *STORE_LOADER_get0_scheme(const STORE_LOADER *store_loader);
 /* struct store_loader_ctx_st is defined differently by each loader */
-typedef struct store_loader_ctx_st STORE_LOADER_CTX;
+typedef void STORE_LOADER_CTX;
 typedef STORE_LOADER_CTX *(*STORE_open_fn)(const STORE_LOADER *loader,
                                            const char *uri,
                                            const UI_METHOD *ui_method,
````
And then:
````
diff --git a/crypto/store/loader_file.c b/crypto/store/loader_file.c
index b7cca5fc0a..fb36c72ee2 100644
--- a/crypto/store/loader_file.c
+++ b/crypto/store/loader_file.c
@@ -594,8 +594,9 @@ struct store_loader_ctx_st {
     } _;
 };
 
-static void STORE_LOADER_CTX_free(STORE_LOADER_CTX *ctx)
+static void STORE_LOADER_CTX_free(STORE_LOADER_CTX *ctxin)
 {
+    struct store_loader_ctx_st *ctx = (struct store_loader_ctx_st *)ctxin;
     if (ctx->type == is_dir) {
         OPENSSL_free(ctx->_.dir.uri);
     } else {
````
etc
This seems a lot less intrusive.